### ISSUE TITLE
Expression evaluate issues

### ISSUE THREAD

**Issue by slaviczavik** [maplibre/maplibre-style-spec#500]
**Describe the bug**
Valid expression syntax does not select all objects and throws a warning.

**To Reproduce**
[This simple example](https://jsbin.com/togimukixi/edit?html,output) shows two layers:

first layer with red stroked circles has applied this filter with deprecated syntax:
```json
[
  "any",
  [">", "level", 0],
  ["!has", "level"]
]
```

and second and the last layer with black circles has applied this filter:
```json
[
  "any",
  [">", ["get", "level"], 0],
  ["!", ["has", "level"]]
]
```

These filters should be evaluated equally and should select same objects, but as you can see, the second filter selects just few objects and it throws this warning: **_Expected value to be of type number, but found null instead_**.

**Expected behavior**
The expression syntax should select same objects as the deprecated syntax.

**Screenshots**
<img width="780" alt="Screenshot 2024-01-17 at 9 59 05" src="https://github.com/maplibre/maplibre-style-spec/assets/6020945/a060eadc-79bf-42a0-9455-40397fd6b7d7">

Red stroked circles are those objects with not defined `level` feature key and should be selected via the `["!", ["has", "level"]]` expression.

**Desktop (please complete the following information):**
 - OS: macOS 14.2.1
 - Browser: Chrome
 - Version: 120.0.6099.216

**Additional context**
For some reason the `!` is not applied and only objects with defined `level` feature key are selected.

**Comment by HarelM** [comment]
Thanks for taking the time to report this issue!
Any chance you can simplify the example and have it all "inline" in the same page (i.e. not link to an external style URL)?
It will make things much easier to reason with.
Maybe even create two layer with different color and alpha to show the overlap of the filter?

**Comment by HarelM** [comment]
If I change the order I think I get the same results, can you verify it?
![image](https://github.com/maplibre/maplibre-style-spec/assets/3269297/5c9ab189-3617-424c-a146-cbef1d3d69bf)

I'm not saying it's not a bug, I'm just surprised about the results...

**Comment by HarelM** [comment]
If you'd like to look into it, there are elaborated expression tests in this repo and you should be able to create some tests for this scenario and test it to see if the behavior, once a test is in place I think it might be simple to fix it.

**Comment by slaviczavik** [comment]
> If I change the order I think I get the same results, can you verify it?
> 
> ![image](https://github.com/maplibre/maplibre-style-spec/assets/3269297/5c9ab189-3617-424c-a146-cbef1d3d69bf)
> 
> 
> 
> I'm not saying it's not a bug, I'm just surprised about the results...

This order indeed works as expected, but it hadn't worked few months ago.

**Comment by neodescis** [comment]
My initial guess is that the "deprecated" expression parser is simply being less strict about type checking. I think in both examples it would be preferable to swap the order to check if the value exists first.

**Comment by slaviczavik** [comment]
> My initial guess is that the "deprecated" expression parser is simply being less strict about type checking. I think in both examples it would be preferable to swap the order to check if the value exists first.

Is there a specific reason why it would be better to swap order if `any` operator is used?

**Comment by slaviczavik** [comment]
> If you'd like to look into it, there are elaborated expression tests in this repo and you should be able to create some tests for this scenario and test it to see if the behavior, once a test is in place I think it might be simple to fix it.

I have just tested the expression with following test and it pass as expected (same result for flipped order):

```
{
  "expression": [
    "any",
    [">", ["get", "level"], 0],
    ["!", ["has", "level"]]
  ],
  "inputs": [
    [{}, {"properties": {"level": 5}}],
    [{}, {"properties": {"level": 0}}],
    [{}, {"properties": {"level": -5}}],
    [{}, {"properties": {"level": null}}]
  ],
  "expected": {
    "compiled": {
      "result": "success",
      "isFeatureConstant": false,
      "isZoomConstant": true,
      "type": "boolean"
    },
    "outputs": [
      true,
      false,
      false,
      {"error":"Expected value to be of type number, but found null instead."}
    ]
  }
}
```

**Comment by neodescis** [comment]
> > My initial guess is that the "deprecated" expression parser is simply being less strict about type checking. I think in both examples it would be preferable to swap the order to check if the value exists first.
> 
> Is there a specific reason why it would be better to swap order if `any` operator is used?

I haven't verified this, but I would expect "any" to use [short-circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation). Basically, no need to keep running through all the checks if only one needs to pass.

**Comment by slaviczavik** [comment]
> > > My initial guess is that the "deprecated" expression parser is simply being less strict about type checking. I think in both examples it would be preferable to swap the order to check if the value exists first.
> > 
> > 
> > Is there a specific reason why it would be better to swap order if `any` operator is used?
> 
> I haven't verified this, but I would expect "any" to use [short-circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation). Basically, no need to keep running through all the checks if only one needs to pass.

That's true, but first expression from the example with `>` is evaluated to false (because object doesn't contain the attribute) so it continue to the expression with `!` and it should be evaluated to true.

**Comment by neodescis** [comment]
> > > > My initial guess is that the "deprecated" expression parser is simply being less strict about type checking. I think in both examples it would be preferable to swap the order to check if the value exists first.
> > > 
> > > 
> > > Is there a specific reason why it would be better to swap order if `any` operator is used?
> > 
> > 
> > I haven't verified this, but I would expect "any" to use [short-circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation). Basically, no need to keep running through all the checks if only one needs to pass.
> 
> That's true, but first expression from the example with `>` is evaluated to false (because object doesn't contain the attribute) so it continue to the expression with `!` and it should be evaluated to true.
> 
> But yes, it makes sense to put `!` operator first, but validator should accept it even if the negation is not first.

If the value is not there, then it is (apparently) comparing whether `null > 0`. I think it's perfectly acceptable for the validator to throw an error on this check. It is also documented as such:  https://maplibre.org/maplibre-style-spec/expressions/#%3E
